mesh/ballotwriter/ballotwriter.go

```diff
@@ -0,0 +1,206 @@
+				for _, ballot := range batch {
+					if !ballot.IsMalicious() {
+						layerBallotStart := time.Now()
+						prev, err := ballots.LayerBallotByNodeID(tx, ballot.Layer, ballot.SmesherID)
+						if err != nil && !errors.Is(err, sql.ErrNotFound) {
+							return err
+						}
+						layerBallotDur += time.Since(layerBallotStart)
+
+						if prev != nil && prev.ID() != ballot.ID() {
+							var ballotProof wire.BallotProof
+							for i, b := range []*types.Ballot{prev, ballot} {
+								ballotProof.Messages[i] = wire.BallotProofMsg{
+									InnerMsg: types.BallotMetadata{
+										Layer:   b.Layer,
+										MsgHash: types.BytesToHash(b.HashInnerBytes()),
+									},
+									Signature: b.Signature,
+									SmesherID: b.SmesherID,
+								}
+							}
+							proof := &wire.MalfeasanceProof{
+								Layer: ballot.Layer,
+								Proof: wire.Proof{
+									Type: wire.MultipleBallots,
+									Data: &ballotProof,
+								},
+							}
+							encoded := codec.MustEncode(proof)
+							if err := identities.SetMalicious(tx, ballot.SmesherID, encoded, time.Now()); err != nil {
+								return fmt.Errorf("add malfeasance proof: %w", err)
+							}
+							ballot.SetMalicious()
+							w.logger.Warn("smesher produced more than one ballot in the same layer",
+								zap.Stringer("smesher", ballot.SmesherID),
+								zap.Object("prev", prev),
+								zap.Object("curr", ballot),
+							)
+						}
+					}
+					ballotAddStart := time.Now()
+					if err := ballots.Add(tx, ballot); err != nil && !errors.Is(err, sql.ErrObjectExists) {
+						return err
+					}
+					ballotAddDur += time.Since(ballotAddStart)
+
+				}
+				return nil
+			}); err != nil {
+				res.err = err
+				ErroredBatchCount.Inc()
+				w.logger.Error("flush ballots to db", zap.Error(err))
+			}
+			cleanupStart := time.Now()
+			w.ballotBatchResult = &batchResult{doneC: make(chan struct{})}
+			for k := range w.ballotBatchIdentities {
+				w.rmNodeSlice = append(w.rmNodeSlice, k)
+			}
+			for _, k := range w.rmNodeSlice {
+				delete(w.ballotBatchIdentities, k)
+			}
+			w.rmNodeSlice = w.rmNodeSlice[:0]
+
+			for k := range batch {
+				w.rmSlice = append(w.rmSlice, k)
+			}
+			for _, k := range w.rmSlice {
+				delete(batch, k)
+			}
+			w.rmSlice = w.rmSlice[:0]
+			w.atxMu.Unlock()
+			writeTime := time.Since(startTime)
+			WriteTime.Add(float64(writeTime))
+			WriteTimeHist.Observe(writeTime.Seconds())
+			LayerBallotTime.Add(float64(layerBallotDur))
+			BallotAddTime.Add(float64(ballotAddDur))
+			CleanupTime.Add(float64(time.Since(cleanupStart)))
+			close(res.doneC)
+		}
+	}
+}
+
+// Store a ballot. Returns a channel to wait for the result, a function that returns
+// the error value and a wait function to try again. The retry logic can happen when
+// a node has multiple ballots on the same layer (malfeasant node or duplicate ballot received).
+func (w *BallotWriter) Store(b *types.Ballot) (<-chan struct{}, func() error, func()) {
```

Comment:
I think it would be much easier if this method were blocking. It can wait for the batch to be submitted internally, wdyt?
```suggestion
func (w *BallotWriter) Store(ctx context.Context, b *types.Ballot) (error) {
```

Reply:
I have a slight aversion to downstream components having their own retry logic. I generally try to avoid this by policy and allow the calling context to do retries (it also usually just makes reading the code and keeping a mental map of it more difficult). 

It will, however, simplify the retry logic from the calling context; agreed. I am generally happy with it as-is, with the only caveat that I now realize there's a missing select case on the for loop that checks for an expired context before calling `Store` (this, however, is likely non-fatal because it will just queue the value for storing, later on exiting on a context cancelled, unless there's a retry in progress).

Reply:
Making this method blocking doesn't mean that the caller cannot retry. It can decide whether to retry or not based on the returned error. But, why would the caller want to retry? Which failures are retriable?

Reply:
I looked at the code again to understand when the `retry` is used. It's used to avoid an ID putting more than 1 ballot into a batch to "avoid problems with malfeasance detection" (double ballot publish). I think this is not needed. If an ID puts 2 ballots for the same layer in a batch, then the existing malfeasance detection logic will detect it. Precisely, the following logic will catch it:

```go
db.WithTx(context.Background(), func(tx sql.Transaction) error {
	require.NoError(t, ballots.Add(tx, ballot))
	// check if ballot is found
	existing, err := ballots.LayerBallotByNodeID(tx, ballot.Layer, ballot.SmesherID)
	// err == nil
	// existing == ballot
}
```

The following test passes:
```go
func TestMalfeasanceDetectionInBatch(t *testing.T) {
	db := statesql.InMemoryTest(t)
	ballot := types.NewExistingBallot(
		types.BallotID{1, 2, 3},
		types.EmptyEdSignature,
		types.RandomNodeID(),
		77,
	)
	db.WithTx(context.Background(), func(tx sql.Transaction) error {
		require.NoError(t, ballots.Add(tx, &ballot))

		// check if ballot is found
		existing, err := ballots.LayerBallotByNodeID(tx, ballot.Layer, ballot.SmesherID)
		require.NoError(t, err)
		require.Equal(t, ballot, *existing)
		return nil
	})
}
```

Therefore I think the `BallotWriter::Store()` can safely put the ballot into the batch without checking the "dirty" set `w.ballotBatchIdentities`. I suggest to change this method to:

```go
func (w *BallotWriter) Store(ctx context.Context, b *types.Ballot) error {
	w.atxMu.Lock()
	br := w.ballotBatchResult
	w.ballotBatch[b.ID()] = b
	w.atxMu.Unlock()

	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-br.doneC:
		return br.err
	}
}
```

Reply:
Yeah so the idea was that one smesher ID cannot write more than one ballot per batch because I didn't want to add the complexity of checking whether a ballot exists for the smesher in the batch, and if yes, check the ID of the ballot, compare against the other, and generate a malfeasance proof as a result (optionally). I felt this is such a rare case it does not worth the effort of having this amount of extra logic - better just check if an entry exists for the identity in the batch and basically tell the caller to back-off on the request instead, trying again later.
To me there are possibly two different scenarios:
- the same ballot arriving twice by gossip or sync from different peers (I'm not sure if this is possible but I wanted to guard against it anyway) at the same time. this is non-fatal and should be idempotent anyway
- two different ballots from the same peer come at the same time (creating a malfeasance proof)
 - these two end up on the same batch
 - if we don't check the identity limitation (one-per-identity-per-batch), the main question is - if there's two ballots and I write them within the domain of the transaction, when the transaction does a query after "inserting"(but not committing) one of them, will it see that same uncommitted ballot when querying the table, if the transaction is not yet committed?

Reply:
> the main question is - if there's two ballots and I write them within the domain of the transaction, when the transaction does a query after "inserting"(but not committing) one of them, will it see that same uncommitted ballot when querying the table, if the transaction is not yet committed?

My claim is that it will, which the test above shows. The additional check and back off mechanism is not needed. You could modify the retry test to verify this.

Reply:
Ok, point made about not needing to keep track over identities, however, passing the context that you want to pass is incorrect, simply because the forever loop of the write-coalescer has a different context (which is cancelled on shutdown).

The problem here is that if I pass the context which is propagated from the request, we might cancel the context but still end up with the specific ballot being written to the database which I find to be counter-intuitive and probably also leading to some inconsistencies. Would still vote to keep this without passing the context to the `Store` method.